sdk/storage/Azure.Storage.Common/src/AsyncCollection.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Storage;
+
+// TODO: Move these types into the Azure.Core package next release?
+
+namespace Azure
+{
+    /// <summary>
+    /// A collection of values that may take multiple service requests to
+    /// iterate over.
+    /// </summary>
+    /// <typeparam name="T">The type of the values.</typeparam>
+    public abstract class AsyncCollection<T> : IAsyncEnumerable<Response<T>>, IEnumerable<Response<T>>
+    {
+        /// <summary>
+        /// Gets a <see cref="CancellationToken"/> used for requests made while
+        /// enumerating asynchronously.
+        /// </summary>
+        protected virtual CancellationToken CancellationToken { get; private set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating the size of <see cref="Page{T}"/>s
+        /// that should be requested from service operations that support it.
+        /// </summary>
+        public virtual int? PageSizeHint { get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncCollection{T}"/>
+        /// class for mocking.
+        /// </summary>
+        protected AsyncCollection() =>
```

Comment:
super nit: Let's remove this ctor. Everyone should have a cancellation token when creating instances of this type.

Reply:
We don't want this even for mocking?

Reply:
Good point, I wonder if people would want to implement this as a mock, maybe we should just give them static factory methods.


`AsyncCollection.Create(Response<T>[])`
`AsyncCollection.Create(Page<T>[])`